src/main/java/gregtech/common/mui/widget/GTFluidSlot.java

```diff
@@ -0,0 +1,350 @@
+                }
+                replaceCursorItemStack(fluidContainer);
+                playSound(tankFluid, false);
+                return fluidContainer;
+            }
+            return ItemStack.EMPTY;
+        }
+
+        /**
+         * Replace the ItemStack on the player's cursor with the passed stack.
+         * Use to replace empty cells with filled, or filled cells with empty.
+         * If it is not fully emptied/filled, it will place the new items into the player inventory instead,
+         * and shrink the held stack by the appropriate amount.
+         */
+        private void replaceCursorItemStack(ItemStack resultStack) {
+            int resultStackSize = resultStack.getMaxStackSize();
+            ItemStack playerStack = getSyncManager().getCursorItem();
+
+            if (!getSyncManager().isClient())
+                syncToClient(1, buffer -> NetworkUtils.writeItemStack(buffer, resultStack));
+
+            while (resultStack.getCount() > resultStackSize) {
+                playerStack.shrink(resultStackSize);
+                addItemToPlayerInventory(resultStack.splitStack(resultStackSize));
+            }
+            if (playerStack.getCount() == resultStack.getCount()) {
+                // every item on the cursor is mutated, so leave it there
+                getSyncManager().setCursorItem(resultStack);
+            } else {
+                // some items not mutated. Mutated items go into the inventory/world.
+                playerStack.shrink(resultStack.getCount());
+                getSyncManager().setCursorItem(playerStack);
+                addItemToPlayerInventory(resultStack);
+            }
+        }
+
+        /** Place an item into the player's inventory, or drop it in-world as an item entity if it cannot fit. */
+        private void addItemToPlayerInventory(ItemStack stack) {
+            if (stack == null) return;
+            var player = getSyncManager().getPlayer();
+
+            if (!player.inventory.addItemStackToInventory(stack) && !player.world.isRemote) {
+                EntityItem dropItem = player.entityDropItem(stack, 0);
+                if (dropItem != null) dropItem.setPickupDelay(0);
+            }
+        }
+
+        /**
+         * Play the appropriate fluid interaction sound for the fluid.
+         * <br />
+         * Must be called on server to work correctly
+         **/
+        private void playSound(FluidStack fluid, boolean fill) {
+            if (fluid == null) return;
+            SoundEvent soundEvent;
+            if (fill) {
+                soundEvent = fluid.getFluid().getFillSound(fluid);
+            } else {
+                soundEvent = fluid.getFluid().getEmptySound(fluid);
+            }
+            EntityPlayer player = getSyncManager().getPlayer();
+            player.world.playSound(null, player.posX, player.posY + 0.5, player.posZ,
+                    soundEvent, SoundCategory.BLOCKS, 1.0F, 1.0F);
```

Comment:
Is there a better `SoundCategory` for this? Maybe should use whichever one regular button press sounds etc. use.